### PR TITLE
[SPARK-27849][SQL] Redact treeString of FileTable and DataSourceV2ScanExecBase

### PR DESCRIPTION
**Pull request by gengliangwang**
## What changes were proposed in this pull request?

To follow https://github.com/apache/spark/pull/17397, the output of FileTable and DataSourceV2ScanExecBase can contain sensitive information (like Amazon keys). Such information should not end up in logs, or be exposed to non-privileged users.

This PR is to add a redaction facility for these outputs to resolve the issue. A user can enable this by setting a regex in the same spark.redaction.string.regex configuration as V1.
## How was this patch tested?

Unit test

### CODE REVIEW

**Comment by gengliangwang** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala [hunk 1]
```diff
@@ -49,6 +51,13 @@ trait FileDataSourceV2 extends TableProvider with DataSourceRegister {
   }
 
   protected def getTableName(paths: Seq[String]): String = {
-    shortName() + ":" + paths.mkString(";")
+    val name = shortName().capitalize + "Table " + paths.map(qualifiedPathName).mkString(",")
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: This changes the table name from:
```
Orc:/path1;/path2
```
To
```
OrcTable file:/path1,file:/path2
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala [hunk 2]
```diff
@@ -35,7 +36,14 @@ trait DataSourceV2ScanExecBase extends LeafExecNode with ColumnarBatchScan {
   def readerFactory: PartitionReaderFactory
 
   override def simpleString(maxFields: Int): String = {
-    s"$nodeName${truncatedString(output, "[", ", ", "]", maxFields)} ${scan.description()}"
+    redact(s"$nodeName${truncatedString(output, "[", ", ", "]", maxFields)} ${scan.description()}")
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala]: nit: seems we only need to redact `scan.description`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: how about `Orc file:/path1,file:/path2`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 3]
```diff
@@ -30,12 +30,13 @@ class DataSourceScanExecRedactionSuite extends QueryTest with SharedSQLContext {
 
   override protected def sparkConf: SparkConf = super.sparkConf
     .set("spark.redaction.string.regex", "file:/[\\w_]+")
+    .set(SQLConf.USE_V1_SOURCE_READER_LIST.key, "orc")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: why this change?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala (hunk 2 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala]: Well, I think redacting the whole string is safer.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Otherwise, we still need to set the conf to use Parquet V1 after Parquet V2 is migrated.

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala [hunk 4]
```diff
@@ -58,4 +58,8 @@ case class OrcScan(
   }
 
   override def hashCode(): Int = getClass.hashCode()
+
+  override def description(): String = {
+    super.description() + ",PushedFilters: " + pushedFilters.mkString("[", ", ", "]")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/orc/OrcScan.scala]: nit. `,PushedFilters` -> `, PushedFilters`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSQLContext
+
+/**
+ * Suite that tests the redaction of DataSourceScanExec
+ */
+class DataSourceV2ScanExecRedactionSuite extends QueryTest with SharedSQLContext {
+
+  override protected def sparkConf: SparkConf = super.sparkConf
+    .set("spark.redaction.string.regex", "file:/[\\w_/]+")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: nit. `"spark.redaction.string.regex"` -> `STRING_REDACTION_PATTERN.key`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import org.apache.spark.SparkConf
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSQLContext
+
+/**
+ * Suite that tests the redaction of DataSourceScanExec
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: `DataSourceScanExec` -> `DataSourceV2ScanExec`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      assert(df.queryExecution.toString.contains(replacement))
+      assert(df.queryExecution.simpleString.contains(replacement))
+    }
+  }
+
+  private def isIncluded(queryExecution: QueryExecution, msg: String): Boolean = {
+    queryExecution.toString.contains(msg) ||
+      queryExecution.simpleString.contains(msg) ||
+      queryExecution.stringWithStats.contains(msg)
+  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: We already have the same function in `DataSourceScanExecRedactionSuite`. Can we try to reduce the duplications, @cloud-fan and @gengliangwang ?
```scala
  private def isIncluded(queryExecution: QueryExecution, msg: String): Boolean = {
    queryExecution.toString.contains(msg) ||
    queryExecution.simpleString.contains(msg) ||
    queryExecution.stringWithStats.contains(msg)
  }
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        // Respect SQLConf and replace FileScan
+        assert(isIncluded(df.queryExecution, replacement))
+
+        assert(!isIncluded(df.queryExecution, "BatchScan"))
+        assert(isIncluded(df.queryExecution, "file:/"))
+      }
+    }
+  }
+
+  test("FileSourceScanExec metadata") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: Ur, `FileSourceScanExec`? Is this correct?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala (hunk 7 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: I think this is quite minor. We should decide the principle of adding tests for data source V2:
1. Try to make tests for V2 isolated. If there is a lot of common code, create an abstract class for the test. So that someday, when the V1 is deprecated, we can remove the tests easier.

2. Put both V1 and V2 test cases together to reduce duplications.


Personally, I prefer option 1.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala (hunk 8 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: Thanks for the careful review!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceV2ScanExecRedactionSuite.scala]: We had better achieve (1) by putting the utility functions into a proper utility class or we design the class hierarchy more reasonably. Removing V1 is not hard if we have a correct design. Most design principles are for reducing duplications and reusing as much as possible.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Let's move the utility function `isIncluded` to a better place and reuse it.

**Comment by gengliangwang** [comment]
@dongjoon-hyun I think the test suites are much better now. Please review it again, thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for redesigning, @gengliangwang . I'll take a look.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala [hunk 9]
```diff
@@ -49,6 +51,13 @@ trait FileDataSourceV2 extends TableProvider with DataSourceRegister {
   }
 
   protected def getTableName(paths: Seq[String]): String = {
-    shortName() + ":" + paths.mkString(";")
+    val name = shortName().capitalize + " " + paths.map(qualifiedPathName).mkString(",")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: `shortName().capitalize` looks like beyond the initial goal of this PR. Is this capitalization is required to pass the tests? Otherwise, shall we keep the existing behavior?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 10]
```diff
@@ -42,6 +44,21 @@ abstract class FileScan(
     false
   }
 
+  override def description(): String = {
+    def seqToString(seq: Seq[Any]) = seq.mkString("[", ", ", "]")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: It seems that we don't need to declare this one-time function. It's different from `FileSourceScanExec` which reuses this function multiple times.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 11]
```diff
@@ -42,6 +44,21 @@ abstract class FileScan(
...
+    val locationDesc = fileIndex.getClass.getSimpleName + seqToString(fileIndex.rootPaths)
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: `seqToString(fileIndex.rootPaths)` -> `fileIndex.rootPaths.mkString("[", ", ", "]")`. And, this may need new line.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala [hunk 12]
```diff
@@ -42,6 +44,21 @@ abstract class FileScan(
...
+    val metadata: Map[String, String] = Map(
+      "ReadSchema" -> readDataSchema.catalogString,
+      "Location" -> locationDesc)
+    val metadataStr = metadata.toSeq.sorted.map {
+      case (key, value) =>
+        val redactedValue =
+          Utils.redact(sparkSession.sessionState.conf.stringRedactionPattern, value)
+        key + ": " + StringUtils.abbreviate(redactedValue, 100)
+    }.mkString(",")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileScan.scala]: `","` -> `", "`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 13]
```diff
@@ -19,26 +19,35 @@ package org.apache.spark.sql.execution
...
 import org.apache.spark.SparkConf
-import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.{DataFrame, QueryTest}
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSQLContext
 
 /**
  * Suite that tests the redaction of DataSourceScanExec
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Could you update the description, too?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 14]
```diff
@@ -19,26 +19,35 @@ package org.apache.spark.sql.execution
...
  */
-class DataSourceScanExecRedactionSuite extends QueryTest with SharedSQLContext {
+abstract class DataSourceScanRedactionTest extends QueryTest with SharedSQLContext {
 
   override protected def sparkConf: SparkConf = super.sparkConf
-    .set("spark.redaction.string.regex", "file:/[\\w_]+")
+    .set("spark.redaction.string.regex", "file:/[\\w_/]+")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Do we need to change the pattern by adding '/'?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 15]
```diff
@@ -53,18 +62,21 @@ class DataSourceScanExecRedactionSuite extends QueryTest with SharedSQLContext {
       assert(df.queryExecution.simpleString.contains(replacement))
     }
   }
+}
 
-  private def isIncluded(queryExecution: QueryExecution, msg: String): Boolean = {
-    queryExecution.toString.contains(msg) ||
-    queryExecution.simpleString.contains(msg) ||
-    queryExecution.stringWithStats.contains(msg)
-  }
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Could you add a document like https://github.com/apache/spark/pull/24719/files#diff-0c515221ed6e6eadcec71b3b9ad3a3e1R116 ?

**Comment by dongjoon-hyun** [comment]
I left a few comments. The others look good. Thank you for updating.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala (hunk 14 above)
Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Yes I think so. The original test case can pass because of the output is truncated as such form:
```
*********(redacted)/var/folders/ml/1fm8cdqd3v39v56sx8s9v94m0000gp/T/spark-28300...
```
If the temporary path is shorter, the test won't pass.

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105902/testReport/org.apache.spark.sql.execution/DataSourceScanExecRedactionSuite/treeString_is_redacted/
We can that the redacted path is 
```
*********(redacted)@2/target/tmp/spark-4a1fc3c6-9e60-43ab-956b-821ee8a6f803
```
We should include the symbol `-` and `@`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala [hunk 16]
```diff
@@ -49,6 +51,13 @@ trait FileDataSourceV2 extends TableProvider with DataSourceRegister {
   }
 
   protected def getTableName(paths: Seq[String]): String = {
-    shortName() + ":" + paths.mkString(";")
+    val name = shortName() + " " + paths.map(qualifiedPathName).mkString(",")
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: Sorry, it seems that I missed this last night. `","` -> `";"`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 17]
```diff
@@ -19,26 +19,35 @@ package org.apache.spark.sql.execution
...
-import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.{DataFrame, QueryTest}
+import org.apache.spark.sql.execution.datasources.v2.BatchScanExec
+import org.apache.spark.sql.execution.datasources.v2.orc.OrcScan
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.test.SharedSQLContext
 
 /**
- * Suite that tests the redaction of DataSourceScanExec
+ *  Test suite base for testing the redaction of DataSourceScanExec/BatchScanExec.
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: nit. Please remove one redundant space between '*' and 'Test'.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala [hunk 18]
```diff
@@ -19,26 +19,35 @@ package org.apache.spark.sql.execution
...
  */
-class DataSourceScanExecRedactionSuite extends QueryTest with SharedSQLContext {
+abstract class DataSourceScanRedactionTest extends QueryTest with SharedSQLContext {
 
   override protected def sparkConf: SparkConf = super.sparkConf
-    .set("spark.redaction.string.regex", "file:/[\\w_]+")
+    .set("spark.redaction.string.regex", "file:/[\\w-_@/]+")
+
+  protected def isIncluded(queryExecution: QueryExecution, msg: String): Boolean = {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Maybe, `final protected` will be clearer.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala (hunk 16 above)
Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: Oh, this is following the output of V1. I am actually fine with either way. WDYT?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala (hunk 14 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/DataSourceScanExecRedactionSuite.scala]: Thank you so much for the explanation!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala (hunk 16 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: Maybe, I missed the context. Which part are you referring? It seems that we replaced ":" to " ", too. Are both of them changed to be consistent with the output of V1?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FileDataSourceV2.scala]: Yes, I think so.
We have to change the ":" to " ", otherwise the output will be like "orc:file:/path.."

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.
Thank you, @gengliangwang and @cloud-fan !

**Comment by gengliangwang** [comment]
Thanks for the review, @dongjoon-hyun @cloud-fan

**Comment by srowen** [comment]
@gengliangwang Hm, weirdly I'm seeing failures in the master build like...
https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-master-test-maven-hadoop-2.7/6415
```
DataSourceV2ScanExecRedactionSuite:
- treeString is redacted *** FAILED ***
  "== Parsed Logical Plan ==
  RelationV2[a#863554L, foo#863555] orc *********(redacted).7/sql/core/target/tmp/spark-9a065829-75ff-450e-90eb-ec8e344851da
```
The PR builder passed so I'm not sure what happened. Does that suggest anything to you?

**Comment by gengliangwang** [comment]
@srowen I think the dot `.` is not matched.
